raft/fsm.hh

```diff
@@ -396,11 +396,16 @@ public:
         _ping_leader = true;
     }
 
+    struct memory_permit {
+        lw_shared_ptr<seastar::semaphore> sem;
+        semaphore_units<> units;
+    };
```

Comment:
This means the concurrency is unlimited. You create a semaphore, then forget it and all pending work. Later (presumably) you create a new semaphore, and it starts serving more work. This can repeat indefinitely.

I'm not familiar with the code, but it looks to me like the semaphore should live outside `struct leader`, which should then get a reference to it. This way there is exactly one semaphore and it outlives the leader. If you want  to be sure the leader is done, wait for the units to be returned.

Reply:
It is an admission semaphore on the leader. When a node become the leader it creates the semaphore with available units (which depends on the log length). What does it mean "if you want to be sure the leader is done"? The node does not have to wait for the log to become zero to revert to non leader. In fact with certain configs the log will never become zero.

Reply:
Is it okay for the tasks to continue executing after the leader is done?

Reply:
I don't understand what this semaphore protects. But if it's lifetime is decoupled, then how can it protect anything?

Reply:
> Is it okay for the tasks to continue executing after the leader is done?

What tasks? Those that are waiting for the semaphore will be aborted. Those that are already admitted are already done. They only put data into the log for raft leader to redistribute. If the node is no longer a leader it does nothing. The new leader will continue.

Reply:
> I don't understand what this semaphore protects. But if it's lifetime is decoupled, then how can it protect anything?

It is an admission semaphore. It makes sure the the leader does not admit new entries when the log is at max size. When a node become a leader it is initialized with available units. When it stops be a leader it it destroyed.

Reply:
I think that Avi's confusion is that the units are not returned when the task is admitted admitted, but rather they are released. They are replenished in `fsm::apply_snapshot` if I'm not mistaken.

Reply:
> > > Is it okay for the tasks to continue executing after the leader is done?
> > 
> > 
> > What tasks? Those that are waiting for the semaphore will be aborted. Those that are already admitted are already done.
> 
> What's the point of the semaphore then?

To not admit new entries when the log is over the limit.

> 
> > They only put data into the log for raft leader to redistribute. If the node is no longer a leader it does nothing. The new leader will continue.
> 
> So they're not done. They exist somewhere, presumably waiting on I/O. Maybe they check somewhere if the node is still a leader, but they can't do it while waiting.

They exist in the memory. The do not wait for any I/O, the do not check somewhere if the node is still a leader they do not wait at all. fsm does not have code that waits. They do not have any work associated with them if the node is not a leader. If it is it tries to replicate them which is a general task that is not associated with a particular entry.

Reply:
```c++

future<entry_id> server_impl::add_entry_on_leader(command cmd, seastar::abort_source* as) {
    // Wait for sufficient memory to become available
    semaphore_units<> memory_permit;
    try {
        memory_permit = co_await _fsm->wait_for_memory_permit(as, log::memory_usage_of(cmd, _config.max_command_size));
    } catch (semaphore_aborted&) {
        throw request_aborted();
    }
    logger.trace("[{}] adding entry after waiting for memory permit", id());

    try {
        const log_entry& e = _fsm->add_entry(std::move(cmd));
        memory_permit.release();
        co_return entry_id{.term = e.term, .idx = e.idx};
    } catch (const not_a_leader&) {
        // the semaphore is already destroyed, prevent memory_permit from accessing it
        memory_permit.release();
        throw;
    }
}
```

If we yield after getting the permit but before co_await returns, we have an outstanding task. At a minimum the number of outstanding tasks is unbounded. At worst we call _fsm->add_entry() when it isn't prepared for it.

The semaphore should have the same lifetime as what it guards. If you can create semaphores whenever you like, you have unlimited units. Granted becoming a leader is slow, but that's besides the point.

Rust gets this right, its semaphore wraps the guarded resource and so shares its lifetime. https://docs.rs/semaphore/latest/semaphore/struct.Semaphore.html

Reply:
I think the fix is simple, move the semaphore to the server. You even use it in the server!

Reply:
> > > > Is it okay for the tasks to continue executing after the leader is done?
> > > 
> > > 
> > > What tasks? Those that are waiting for the semaphore will be aborted. Those that are already admitted are already done.
> > 
> > 
> > What's the point of the semaphore then?
> 
> To not admit new entries when the log is over the limit.
> 


Then it's not working, since you can have any number of semaphores for the same log.

Reply:
> ```c++
> future<entry_id> server_impl::add_entry_on_leader(command cmd, seastar::abort_source* as) {
>     // Wait for sufficient memory to become available
>     semaphore_units<> memory_permit;
>     try {
>         memory_permit = co_await _fsm->wait_for_memory_permit(as, log::memory_usage_of(cmd, _config.max_command_size));
>     } catch (semaphore_aborted&) {
>         throw request_aborted();
>     }
>     logger.trace("[{}] adding entry after waiting for memory permit", id());
> 
>     try {
>         const log_entry& e = _fsm->add_entry(std::move(cmd));
>         memory_permit.release();
>         co_return entry_id{.term = e.term, .idx = e.idx};
>     } catch (const not_a_leader&) {
>         // the semaphore is already destroyed, prevent memory_permit from accessing it
>         memory_permit.release();
>         throw;
>     }
> }
> ```
> 
> If we yield after getting the permit but before co_await returns, we have an outstanding task. At a minimum the number of outstanding tasks is unbounded. At worst we call _fsm->add_entry() when it isn't prepared for it.

add_entry() is always prepared. So the scenario you see is that if we yield after wait_for_memory_permit() a leader my change and then going back to this node and then the semaphore will be recreated and we will proceed with old units. Good catch and easy to fix. We can detect the case by checking the term before and after the yield.

Reply:
You don't know if _fsm even exists. The coroutine can be waiting indefinitely and execute after the whole thing was torn down.

Of course, it's unrealistic in practice, but it's asking for trouble to decouple lifetimes.

Reply:
> > I think the fix is simple, move the semaphore to the server. You even use it in the server!
> 
> Only the leader uses it. The log length may change while a node is not a leader and the semaphore will hold an outdated value.

Then add or remove units as needed.

Reply:
Another alternative is that whoever tears down the leader, first asks it to collect any pending operations.

Reply:
> You don't know if _fsm even exists. The coroutine can be waiting indefinitely and execute after the whole thing was torn down.

The torn down will wait for the code to complete or we need to check that the server is aborted after the sleep. The _fsm is accessed unrelated to my fix here.

> 
> Of course, it's unrealistic in practice, but it's asking for trouble to decouple lifetimes.

Do you see any other problem with the code after the fix?

Reply:
> Another alternative is that whoever tears down the leader, first asks it to collect any pending operations.

Why leader? Whoever tears down the server. And this is the case. The server cannot be destroyed before all outstanding async operation on it completed. So if a use called `raft.add_entry_on_leader()` it cannot destroy `raft` object before the call resolves. This is not different from any other seatstar object.

Reply:
> > > I think the fix is simple, move the semaphore to the server. You even use it in the server!
> > 
> > 
> > Only the leader uses it. The log length may change while a node is not a leader and the semaphore will hold an outdated value.
> 
> Then add or remove units as needed.

That will require to have access to the semaphore in places we do not have it now. Only fsm (consensus module it should have been called) logic knows how many entries were added and it is decoupled from the server by design. And even if we would have adjusted the semaphore units somehow it would not have solved the race you pointed above, so the patch I posted is needed anyway.

Reply:
> > You don't know if _fsm even exists. The coroutine can be waiting indefinitely and execute after the whole thing was torn down.
> 
> The torn down will wait for the code to complete or we need to check that the server is aborted after the sleep. The _fsm is accessed unrelated to my fix here.
> 
> > 
> > Of course, it's unrealistic in practice, but it's asking for trouble to decouple lifetimes.
> 
> Do you see any other problem with the code after the fix?

Yes, it still creates unbounded tasks, then adds workarounds for them.

Reply:
> > > You don't know if _fsm even exists. The coroutine can be waiting indefinitely and execute after the whole thing was torn down.
> > 
> > 
> > The torn down will wait for the code to complete or we need to check that the server is aborted after the sleep. The _fsm is accessed unrelated to my fix here.
> > > Of course, it's unrealistic in practice, but it's asking for trouble to decouple lifetimes.
> > 
> > 
> > Do you see any other problem with the code after the fix?
> 
> Yes, it still creates unbounded tasks, then adds workarounds for them.

A semaphore is not a tool to prevent unbounded tasks. The purpose of the semaphore here is not to prevent unbounded tasks. A semaphore may only prevent unbounded tasks from a single client and it will do it here. Moving the semaphore into another place will not prevent unbounded tasks.